### PR TITLE
Add `--mir-linker` flag and wire it up to the compiler

### LINKED ISSUE #1617: Add `--mir-linker` option to `kani-driver`
This issue captures the work to wire up the new cross-crate reachability analysis and the new sysroot in order to implement the MIR Linker. Until this feature is stabilized, the new flow will be available via the unstable flag `--mir-linker`.

### PR DESCRIPTION
**Pull request by celinval**
### Description of changes: 

This change adds the `--mir-linker` option to `kani-driver` and `--reachability=[none | legacy | harnesses]` to the compiler. I also changed `cargo kani` to invoke `cargo rustc` instead of `cargo [build | tests]` so the reachability flag is passed at the right moment.

### Resolved issues:

Resolves #1617

### Related RFC:

#1600

### Call-outs:

1. Since the MIR linker hasn't been hooked up yet, the compiler will fail if this option is selected as shown by the new tests.
2. `cargo rustc -- --reachability=harnesses` will inject the argument `--reachability=harness` in the middle of the compiler flags. Thus, it has to be extracted from rustc args. :(

### Testing:

* How is this change tested?

* Is this a refactor change?

### Checklist
- [x] Each commit message has a non-empty body, explaining why the change was made
- [x] Methods or procedures are documented
- [x] Regression or unit tests are included, or existing tests cover the modified code
- [x] My PR is restricted to a single feature or bugfix

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 and MIT licenses.

### CODE REVIEW

File: tests/cargo-ui/mir-linker/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+[package]
+name = "mir-linker"
+version = "0.1.0"
+edition = "2021"
+
+[dependencies]
+rand = "0.8.4"
+
+[package.metadata.kani]
+flags = { mir-linker=true, enable-unstable=true }
```
Comment (tedinski) [inline tests/cargo-ui/mir-linker/Cargo.toml:12]: No expected file for this one?

File: kani-driver/src/call_cargo.rs [hunk 2]
```diff
@@ -22,66 +22,67 @@ pub struct CargoOutputs {
...
+            kani_args.push("--reachability=legacy".into());
         }
 
-        let mut cmd = Command::new("cargo");
-        cmd.args(args)
-            .env("RUSTC", &self.kani_compiler)
-            .env("RUSTFLAGS", "--kani-flags")
-            .env("KANIFLAGS", flag_env);
+        // Only joing them at the end. All kani flags must come first.
+        kani_args.append(&mut rustc_args);
```
Comment (tedinski) [inline kani-driver/src/call_cargo.rs:72]: &mut? oh, wait this says append not extend. IMO `extend` is cleaner because it avoids needing the mut

Comment (celinval) [reply kani-driver/src/call_cargo.rs:72]: Sure. I'll use `extend_from_slice()`.

File: tests/cargo-ui/mir-linker/Cargo.toml (hunk 1 above)
Comment (celinval) [reply tests/cargo-ui/mir-linker/Cargo.toml:12]: Ooops. Thanks for the call out.

**Review by tedinski — APPROVED** [review]

**Pull request merged**